### PR TITLE
[flang][OpenMP] Add basic support to lower `loop` directive to MLIR

### PR DESCRIPTION
**Pull request by ergawy**
Adds initial support for lowering the `loop` directive to MLIR.

The PR includes basic suport and testing for the following clauses:
 * `collapse`
 * `order`
 * `private`
 * `reduction`

Parent PR: #113911, only the latest commit is relevant to this PR.

### CODE REVIEW

File: mlir/lib/Conversion/OpenMPToLLVM/OpenMPToLLVM.cpp
Comment (skatrak) [inline mlir/lib/Conversion/OpenMPToLLVM/OpenMPToLLVM.cpp:287]: Ultra-nit: Alphabetical sorting.

File: mlir/lib/Conversion/OpenMPToLLVM/OpenMPToLLVM.cpp
Comment (skatrak) [inline mlir/lib/Conversion/OpenMPToLLVM/OpenMPToLLVM.cpp:241]: Ultra-nit: Alphabetical sorting.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:2037]: Nit: Move this to the corresponding section of the file where other `genXyzClauses` definitions are located.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:2045]: Is `lastprivate` not already handled by the call in `createBodyOfOp()` to `DataSharingProcessor::processStep2()`?

File: flang/test/Lower/OpenMP/loop-directive.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/loop-directive.f90:48]: Check that the private copy is the one being accessed in the body.

File: flang/test/Lower/OpenMP/loop-directive.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/loop-directive.f90:76]: Check that the reduction block argument is the one being accessed in the body.

**Review by skatrak — COMMENTED** [review]
Thanks Kareem. This mostly LGTM, just small comments.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (ergawy) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2062]: Not out-of-the-box. We hit this TODO: https://github.com/llvm/llvm-project/blob/main/flang/lib/Lower/OpenMP/DataSharingProcessor.cpp#L260. If you don't mind, I prefer to handle this in its own PR to focus on testing `lastprivate`.

**Comment by ergawy** [comment]
Ping 🔔!

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:2066]: Ultra-nit: This doesn't belong in this section of the file, since the `loop` construct never belongs to a composite construct. It should go below `genLoopNestOp`.

File: flang/test/Lower/OpenMP/loop-directive.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/loop-directive.f90:15]: Nit: Use CHECK-NEXT here and in the other instances of `omp.loop_nest` below, since we want to make sure nothing appears between `omp.loop` and `omp.loop_nest`, as it's a loop wrapper.

**Review by skatrak — APPROVED** [review]
Thank you @ergawy, this LGTM. Just very minor nits, but no need for another review by me before merging.

**Pull request merged**
